generate/generate.go

```diff
@@ -134,19 +134,91 @@ func (g *Generator) GeneratePkg(path string, gens []config.Generator) error {
 	}
 	// Run any other configured generators.
 	for _, gen := range gens {
-		if err := g.generatePlugin(*req, gen); err != nil {
-			return err
+		if gen.IsProtoc() {
+			if err := g.generateProtoc(*req, gen); err != nil {
+				return err
+			}
+		} else {
+			if err := g.generatePlugin(*req, gen); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
-func (g *Generator) generatePlugin(req plugin.CodeGeneratorRequest, gen config.Generator) error {
-	protoParameters := make([]string, len(gen.Params))
-	for i, p := range gen.Params {
-		protoParameters[i] = fmt.Sprintf("%s=%s", p.Key, p.Value)
+func (g *Generator) generateProtoc(req plugin.CodeGeneratorRequest, gen config.Generator) error {
+	// The proto files we are asking protoc to generate. This should be
+	// a formatted list of what is in req.FileToGenerate.
+	protoFilenames := []string{}
+
+	fds := desc.FileDescriptorSet{
+		File: req.ProtoFile,
+	}
+	// Determine the files that protoc will be generating and
+	// remove the package directory path and change .gunk to .proto.
+	// This is needed because there is no way to specify the output
+	// filename, and protoc will use the FileDescriptorSet.File.Name
+	// to determine the filename. In our case, that would be the
+	// package path, but it would be created from the current relative
+	// directory, rather than an absolute path.
+	//
+	// We also replace .gunk with .proto. protoc will turn
+	// 'echo.gunk' into 'echo.gunk_pb.js' which makes it a bit
+	// hard to replace afterwards. This seemed like the easier approach.
+	for i, f := range fds.File {
+		for _, ftg := range req.GetFileToGenerate() {
+			if f.GetName() == ftg {
```

Comment:
invert and early continue to unindent the rest of the loop body

Reply:
Done.